core/src/main/scala/org/apache/spark/deploy/Client.scala

```diff
@@ -176,6 +202,22 @@ private class ClientEndpoint(
       } else if (!Utils.responseFromBackup(message)) {
         System.exit(-1)
       }
+
+    case DriverStatusResponse(found, state, _, _, _) =>
```

Comment:
It would be better if we could do some refactor on `pollAndReportStatus` in order to reduce some duplicate logic.
For example, we can only call `pollAndReportStatus` here, and remove other invocations in `SubmitDriverResponse`/ `KillDriverResponse`. And, of course, the `pollAndReportStatus`(it also needs a new name) will not poll the status anymore.

Reply:
@Ngone51  Thanks for your feedback. `pollAndReportStatus` is only being used the first time after submitting or killing drivers. I am not sure which is the duplicate logic you are referring to. Also, `pollAndReportStatus` is only polling the driver status and handling the response. If we removing polling from that, what logic should be handled there?

Reply:
> If we removing polling from that, what logic should be handled there?

we use this:

```
forwardMessageThread.scheduleAtFixedRate(() => Utils.tryLogNonFatalError {
          MonitorDriverStatus()
        }, 0, REPORT_DRIVER_STATUS_INTERVAL, TimeUnit.MILLISECONDS)
```
(but the initial delay need to change)

in this way, submitting or killing drivers will still use it only for one time when `waitAppCompletion=false`.

Reply:
Scheduling to monitor driver status is done only in case of submit and not in kill as of now.  So we may need to explicitly send a message to monitor driver status after 5 seconds delay in case of kill.

Reply:
that's why I said we need to change a the delay (e.g. 5s) instead of 0 for both submiting and killing.

Reply:
I agree that we can change the delay to 5 seconds to keep it consistent with current logic. My question is that should we add the following block in `case "kill" =>` as well? 
`        forwardMessageThread.scheduleAtFixedRate(() => Utils.tryLogNonFatalError {
          MonitorDriverStatus()
        }, 0, REPORT_DRIVER_STATUS_INTERVAL, TimeUnit.MILLISECONDS)
`

Reply:
we don't need to add `forwardMessageThread.scheduleAtFixedRate(...)` into any `case` branches but just put it as a global one(just do what you do now). I think it still works for `case "kill"`.

Reply:
It's in the case "launch" as of now. I will move it to a global place and refactor the code. Thanks for your suggestions.

Reply:
oh, sorry miss that. yea, thank you!

Reply:
@Ngone51 I have refactored the code as suggested. Kindly review it again. Thanks.